File: exp-10/q2-dfs/solution.py
import sys

sys.setrecursionlimit(1000000)


def dfs(u):
    visited[u] = True
    order.append(u)
    for v in adj[u]:
        if not visited[v]:
            dfs(v)


n, m = map(int, input().split())
adj = [[] for _ in range(n + 1)]
for _ in range(m):
    u, v = map(int, input().split())
    adj[u].append(v)
    adj[v].append(u)
s = int(input())

for i in range(1, n + 1):
    adj[i].sort()

visited = [False] * (n + 1)
order = []
dfs(s)
print("DFS Traversal:", " ".join(map(str, order)))

File: exp-10/q2-dfs/solution.go
package main

import (
	"fmt"
	"sort"
)

var adj [][]int
var visited []bool
var order []int

func dfs(u int) {
	visited[u] = true
	order = append(order, u)
	for _, v := range adj[u] {
		if !visited[v] {
			dfs(v)
		}
	}
}

func main() {
	var n, m int
	fmt.Scan(&n, &m)
	adj = make([][]int, n+1)
	for i := 0; i < m; i++ {
		var u, v int
		fmt.Scan(&u, &v)
		adj[u] = append(adj[u], v)
		adj[v] = append(adj[v], u)
	}
	var s int
	fmt.Scan(&s)

	for i := 1; i <= n; i++ {
		sort.Ints(adj[i])
	}

	visited = make([]bool, n+1)
	dfs(s)
	fmt.Printf("DFS Traversal: ")
	for i, u := range order {
		if i > 0 {
			fmt.Printf(" ")
		}
		fmt.Printf("%d", u)
	}
	fmt.Println()
}